README
Copyright (C) 2010-2022 Jason Woodward <woodwardj at jaos dot org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Library General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


Table of Contents:
 1. About
 2. Features
 3. Requirements
 4. Installation
 5. slapt-srcrc
 6. SLACKBUILDS.TXT
 7. Troubleshooting
 8. Thanks and contributions


1. About

slapt-src is a utility to make querying, retrieving, and building slackbuilds as
easy as working with binary packages with slapt-get.


2. Features

  * Remote and local SlackBuild repository support
  * fast
  * more

3. Requirements

  * slapt-get
  * curl

4. Installation

    You can install via meson or autotools (deprecated).

    * Meson

        meson setup build # or any dir of your choosing
        meson compile -C build # or `ninja -C build` for older meson releases
        meson install -C build # or `ninja -C build install`  for older meson releases

        Meson/ninja supports `DESTDIR=/path meson install -C build`

        TIP:
        slapt-src.Slackbuild will create a Slackware package.
        Run with sudo or as a privileged user.

    * autotools (deprecated)

        ./autogen.sh # if building from git
        ./configure
        make
        make install # as privileged user

        The Makefile honors DESTDIR.

        TIP:
        `make pkg` (as privileged user) will create a Slackware package.

5. slapt-srcrc

The slapt-srcrc file consists of the following keys:
  * SOURCE
  * BUILDDIR
  * PKGEXT
  * PKGTAG

Refer to the slapt-src(8) man page for more on the configuration format and options.

Source should point to a local or remote SlackBuild repository that contains
the necessary meta data.  See SLACKBUILDS.TXT below.

6. SLACKBUILDS.TXT

The following script can generate the currently supported format of SLACKBUILDS.TXT.
The filename and format are likely to change.

#!/bin/bash
set -e

for sb in $(find . -name '*.SlackBuild' | sort)
do
  name=$(basename $sb | sed -re 's/\.SlackBuild$//')
  location=$(dirname $sb)
  if [ -f $location/$name.info ]; then
    echo "SLACKBUILD NAME: $name"
    echo "SLACKBUILD LOCATION: $location"
    files=$(cd $location && find .  -type f -printf '%P\n' | sort | xargs)
    echo "SLACKBUILD FILES: $files"

    # remove those pesky multi line listings for each interesting field
    TMP=$(mktemp)
    sed ':a;N;$!ba;s/\\\n*\s*//g' $location/$name.info > $TMP

    DOWNLOAD=$(grep ^DOWNLOAD= $TMP | cut -f2 -d\" )
    DOWNLOAD_x86_64=$(grep ^DOWNLOAD_x86_64= $TMP | cut -f2 -d\" )
    MD5SUM=$(grep ^MD5SUM= $TMP | cut -f2 -d\" )
    MD5SUM_x86_64=$(grep ^MD5SUM_x86_64= $TMP | cut -f2 -d\" )
    VERSION=$(grep ^VERSION= $TMP | cut -f2 -d\" )

    echo "SLACKBUILD VERSION: $VERSION"
    echo "SLACKBUILD DOWNLOAD: $DOWNLOAD"
    echo "SLACKBUILD DOWNLOAD_x86_64: $DOWNLOAD_x86_64"
    echo "SLACKBUILD MD5SUM: $MD5SUM"
    echo "SLACKBUILD MD5SUM_x86_64: $MD5SUM_x86_64"

    if [ -f $location/slack-desc ]; then
      SHORTDESC=$(grep ^$name: $location/slack-desc | head -n 1 | sed -re "s/^$name://")
      echo "SLACKBUILD SHORT DESCRIPTION: $SHORTDESC"
    else
      echo "SLACKBUILD SHORT DESCRIPTION: "
    fi

    echo
    rm -f $TMP
  fi

done > SLACKBUILDS.TXT
gzip -9 SLACKBUILDS.TXT -c > SLACKBUILDS.TXT.gz

# END

7. Troubleshooting

There are likely a lot of bugs.  Email bug reports to me:
    Jason Woodward <woodwardj at jaos dot org>

Or create issues on Github:
    https://github.com/jaos/slapt-src/issues

8. Thanks and contributions

Thanks to the guys from #gsb on freenode for ideas and conceptual discussions.

Thanks to Darren 'Tadgy' Austin and the sponsors File-away Limited
(http://www.file-away.co.uk) and UK2.NET (http://www.uk2.net) for
mirroring slackbuilds.org.


ChangeLog
0.3.7 Feb 05, 2022
 * tx pull

0.3.6 Mar 12, 2021
 * misc symbol cleanup
 * build updates
 * doc updates

0.3.5 Feb 13, 2021
 * meson build support
 * translation updates

0.3.4 Jul 28, 2019
 * update naming convention for data structures
 * tx pull
 * bump to version 0.3.4
 * move CFLAG customization until after AC_PROG_CC_C99
 * port to slapt-get 0.11.1
 * add PKGREL automake variable for packaging release tag

0.3.3 Jun 15, 2019
 * add slack-suggests with development packages
 * fix for dependency ordering regression
 * use %m and no -pedantic for c99 sscanf
 * build fixes
 * tx pull
 * use c99 stdbool (libslapt compat)

0.3.2i Jul 04, 2016
 * Slackware-arm support
 * translation updates

0.3.2h May  07, 2015
 * add -t|--simulate option
 * skip non-deps (like %README%)

0.3.2g Jan  12, 2015
 * support space delimited dependency list

0.3.2f Apr  05, 2014
 * slapt-get build compatibility
 * translation updates
 * tweak search to return exact name matches in case of regex characters

0.3.2e Apr  29, 2013
 * Updated translations

0.3.2d Sep  03, 2012
 * add explicit openssl autoconf check

0.3.2c Apr  23, 2012
 * updated zlib usage to fix "expected 'gzFile' but argument is of type
   'struct gzFile_s **'".  Thanks Igor
 * drop extra version comparison in finding the generated package

0.3.2b Oct  05, 2011
 * --build name:version no longer auto installs

0.3.2a Aug  02, 2011
 * --show now recognizes the name:version notation
 * updated --help output (thanks Igor)

0.3.2 Jun  12, 2011
 * make sure to accept short hand --clean (-e) and short hand --config (-c)
 * mention --help and --version in the manual
 * SLAPTSRC_SLKBUILD_CMD_LEN too short when using SLKBUILDS
 * removed unimplemented function declaration
 * SLACKBUILD_REQUIRES is comma delimited
 * close builddir after clean
 * updated slapt-srcrc for 13.37 sources

0.3.1a Jan  06, 2011
* Igor reformatted the slapt-src man page

0.3.1 Dec  18, 2010
 * support download resume for the slackbuild downloads
 * return non-zero on faulty command line argument or failed cache download
 * hungarian translation by Peter Polonkai

0.3 Nov  04, 2010
 * updated ru translation from Igor
 * updated Greek translation from George Vlahavas
 * added --clean option
 * some files contain paths, create those paths as necessary
 * es and es_AR translation from seacat

0.2a Oct  29, 2010
 * updated de translation from Manfred
 * updated error handling when parsing configuration
 * updated greek translation from George

0.2 Oct  19, 2010
 * added --postprocess option
 * added de translation from Manfred
 * added fr translation from Cyrille
 * added ru translation from Igor
 * es and es_AR translation from seacat
 * Greek translation from George Vlahavas

0.1 Oct  03, 2010
 * initial release


t/clitests.sh
#!/bin/bash
set -eoxu

TEST_TMPDIR=$(mktemp -d)
trap "rm -rf ${TEST_TMPDIR};" err exit

slaptsrc="${1}"
config=${TEST_TMPDIR}/config
cat > ${config} << EOF
SOURCE=http://www.slackbuilds.org/slackbuilds/14.2/
BUILDDIR=${TEST_TMPDIR}/slapt-src
PKGEXT=txz
EOF

${slaptsrc} --config "${config}" --help
${slaptsrc} --config "${config}" --version
${slaptsrc} --config "${config}" --update
${slaptsrc} --config "${config}" --list
${slaptsrc} --config "${config}" --search test
${slaptsrc} --config "${config}" --show z
${slaptsrc} --config "${config}" --upgrade-all -t
${slaptsrc} --config "${config}" --fetch z -t
${slaptsrc} --config "${config}" --build z -t
${slaptsrc} --config "${config}" --install z -t
${slaptsrc} --config "${config}" --fetch z -y
${slaptsrc} --config "${config}" --clean

find ${TEST_TMPDIR}


src/source.h
/*
 * Copyright (C) 2010-2022 Jason Woodward <woodwardj at jaos dot org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <slapt.h>
#include <sys/utsname.h>
#ifndef __SLAPT_SRC_SOURCE_H__
#define __SLAPT_SRC_SOURCE_H__

#define SLAPT_SRC_RC "/etc/slapt-get/slapt-srcrc"
#define SLAPT_SRC_DATA_FILE "slackbuilds_data"
#define SLAPT_SRC_SOURCE_TOKEN "SOURCE="
#define SLAPT_SRC_BUILDDIR_TOKEN "BUILDDIR="
#define SLAPT_SRC_PKGEXT_TOKEN "PKGEXT="
#define SLAPT_SRC_PKGTAG_TOKEN "PKGTAG="
#define SLAPT_SRC_SOURCES_LIST_GZ "SLACKBUILDS.TXT.gz"
#define SLAPT_SRC_SOURCES_LIST "SLACKBUILDS.TXT"

typedef struct _slapt_src_config_ {
    slapt_vector_t *sources;
    char *builddir;
    char *pkgext;
    char *pkgtag;
    char *postcmd;
    bool do_dep;
    bool prompt;
} slapt_src_config;
slapt_src_config *slapt_src_config_init(void);
void slapt_src_config_free(slapt_src_config *config);
slapt_src_config *slapt_src_read_config(const char *filename);

typedef struct _slapt_src_slackbuild_ {
    char *name;
    char *version;
    char *location;
    char *sb_source_url;
    slapt_vector_t *files;
    char *download;
    char *download_x86_64;
    char *md5sum;
    char *md5sum_x86_64;
    char *short_desc;
    char *requires;
} slapt_src_slackbuild;
slapt_src_slackbuild *slapt_src_slackbuild_init(void);
void slapt_src_slackbuild_free(slapt_src_slackbuild *);

bool slapt_src_update_slackbuild_cache(const slapt_src_config *);
slapt_vector_t *slapt_src_get_available_slackbuilds(void);
bool slapt_src_fetch_slackbuild(const slapt_src_config *, const slapt_src_slackbuild *);
bool slapt_src_build_slackbuild(const slapt_src_config *, const slapt_src_slackbuild *);
bool slapt_src_install_slackbuild(const slapt_src_config *, const slapt_src_slackbuild *);
slapt_vector_t *slapt_src_names_to_slackbuilds(const slapt_src_config *, const slapt_vector_t *, const slapt_vector_t *, const slapt_vector_t *);
slapt_vector_t *slapt_src_get_slackbuilds_from_file(const char *);
void slapt_src_write_slackbuilds_to_file(slapt_vector_t *, const char *);
slapt_vector_t *slapt_src_search_slackbuild_cache(const slapt_vector_t *, const slapt_vector_t *);
slapt_src_slackbuild *slapt_src_get_slackbuild(const slapt_vector_t *, const char *, const char *);

int sb_compare_name_to_name(const void *a, const void *b);
int sb_compare_pkg_to_name(const void *a, const void *b);

#endif


src/main.c
/*
 * Copyright (C) 2010-2022 Jason Woodward <woodwardj at jaos dot org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <slapt.h>
#include <sys/utsname.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <config.h>
#include "source.h"

#define BUILD_ONLY_FLAG 1
#define FETCH_ONLY_FLAG 2

/* getopt */
extern char *optarg;
extern int optind, opterr, optopt;

static int show_summary(slapt_vector_t *, slapt_vector_t *, int, bool);
static void clean(slapt_src_config *config);

void version(void)
{
    printf(gettext("%s version %s\n"), PACKAGE, VERSION);
    printf("Jason Woodward <woodwardj at jaos dot org>\n");
    printf("\n");
    printf("This program is free software; you can redistribute it and/or modify\n");
    printf("it under the terms of the GNU General Public License as published by\n");
    printf("the Free Software Foundation; either version 2 of the License, or\n");
    printf("any later version.\n");
    printf("This program is distributed in the hope that it will be useful,\n");
    printf("\n");
    printf("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
    printf("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
    printf("GNU Library General Public License for more details.\n");
    printf("\n");
    printf("You should have received a copy of the GNU General Public License\n");
    printf("along with this program; if not, write to the Free Software\n");
    printf("Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.\n");
}

void help(void)
{
    printf(gettext("%s - A SlackBuild utility\n"), PACKAGE);
    printf(gettext("Usage: %s [option(s)] [action] [slackbuild(s)]\n"), PACKAGE);
    printf("  -u, --update           %s\n", gettext("update local cache of remote slackbuilds"));
    printf("  -U, --upgrade-all      %s\n", gettext("upgrade all installed slackbuilds"));
    printf("  -l, --list             %s\n", gettext("list available slackbuilds"));
    printf("  -e, --clean            %s\n", gettext("clean build directory"));
    printf("  -s, --search           %s\n", gettext("search available slackbuilds"));
    printf("  -w, --show             %s\n", gettext("show specified slackbuilds"));
    printf("  -i, --install          %s\n", gettext("fetch, build, and install the specified slackbuild(s)"));
    printf("  -b, --build            %s\n", gettext("only fetch and build the specified slackbuild(s)"));
    printf("  -f, --fetch            %s\n", gettext("only fetch the specified slackbuild(s)"));
    printf("  -v, --version\n");
    printf("  -h, --help\n");
    printf(" %s:\n", gettext("Options"));
    printf("  -y, --yes              %s\n", gettext("do not prompt"));
    printf("  -t, --simulate         %s\n", gettext("show what will be done"));
    printf("  -c, --config=FILE      %s\n", gettext("use the specified configuration file"));
    printf("  -n, --no-dep           %s\n", gettext("do not look for dependencies"));
    printf("  -p, --postprocess=CMD  %s\n", gettext("run specified command on generated package"));
    printf("  -B, --build-only       %s\n", gettext("applicable only to --upgrade-all"));
    printf("  -F, --fetch-only       %s\n", gettext("applicable only to --upgrade-all"));
}

#define VERSION_OPT 'v'
#define HELP_OPT 'h'
#define UPDATE_OPT 'u'
#define UPGRADE_OPT 'U'
#define LIST_OPT 'l'
#define SEARCH_OPT 's'
#define SHOW_OPT 'w'
#define FETCH_OPT 'f'
#define BUILD_OPT 'b'
#define INSTALL_OPT 'i'
#define YES_OPT 'y'
#define CONFIG_OPT 'c'
#define NODEP_OPT 'n'
#define POSTCMD_OPT 'p'
#define CLEAN_OPT 'e'
#define SIMULATE_OPT 't'
#define BUILD_ONLY_OPT 'B'
#define FETCH_ONLY_OPT 'F'

struct utsname uname_v; /* for .machine */

static void init_builddir(slapt_src_config *config)
{
    DIR *builddir = NULL;
    mode_t mode = S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;

    if ((builddir = opendir(config->builddir)) == NULL) {
        if (mkdir(config->builddir, mode) == -1) {
            printf(gettext("Failed to create build directory [%s]\n"), config->builddir);

            if (errno)
                perror(config->builddir);

            exit(EXIT_FAILURE);
        }
    } else {
        closedir(builddir);
    }
}

int main(int argc, char *argv[])
{
    static struct option long_options[] = {
        {"build", required_argument, 0, BUILD_OPT},
        {"b", required_argument, 0, BUILD_OPT},
        {"build-only", no_argument, 0, BUILD_ONLY_OPT},
        {"B", no_argument, 0, BUILD_ONLY_OPT},
        {"clean", no_argument, 0, CLEAN_OPT},
        {"e", no_argument, 0, CLEAN_OPT},
        {"config", required_argument, 0, CONFIG_OPT},
        {"c", required_argument, 0, CONFIG_OPT},
        {"fetch", required_argument, 0, FETCH_OPT},
        {"f", required_argument, 0, FETCH_OPT},
        {"fetch-only", no_argument, 0, FETCH_ONLY_OPT},
        {"F", no_argument, 0, FETCH_ONLY_OPT},
        {"help", no_argument, 0, HELP_OPT},
        {"install", required_argument, 0, INSTALL_OPT},
        {"list", no_argument, 0, LIST_OPT},
        {"no-dep", no_argument, 0, NODEP_OPT},
        {"postprocess", required_argument, 0, POSTCMD_OPT},
        {"search", required_argument, 0, SEARCH_OPT},
        {"s", required_argument, 0, SEARCH_OPT},
        {"show", required_argument, 0, SHOW_OPT},
        {"w", required_argument, 0, SHOW_OPT},
        {"simulate", no_argument, 0, SIMULATE_OPT},
        {"t", no_argument, 0, SIMULATE_OPT},
        {"update", no_argument, 0, UPDATE_OPT},
        {"u", no_argument, 0, UPDATE_OPT},
        {"upgrade-all", no_argument, 0, UPGRADE_OPT},
        {"U", no_argument, 0, UPGRADE_OPT},
        {"yes", no_argument, 0, YES_OPT},
        {"version", no_argument, 0, VERSION_OPT},
        {0, 0, 0, 0}};

    /* initialization */
    setbuf(stdout, NULL);
#ifdef ENABLE_NLS
    setlocale(LC_ALL, "");
    textdomain(GETTEXT_PACKAGE);
#endif
#ifdef SLAPT_HAS_GPGME
    gpgme_check_version(NULL);
#ifdef ENABLE_NLS
    gpgme_set_locale(NULL, LC_CTYPE, setlocale(LC_CTYPE, NULL));
#endif
#endif
    curl_global_init(CURL_GLOBAL_ALL);
    uname(&uname_v);

    /* stop early */
    if (argc < 2) {
        help();
        exit(EXIT_SUCCESS);
    }

    int only_flags = 0;
    bool prompt = true, do_dep = true, simulate = false;
    char *config_file = NULL, *postcmd = NULL;
    slapt_vector_t *names = slapt_vector_t_init(free);
    int c = -1, option_index = 0, action = 0;
    while ((c = getopt_long_only(argc, argv, "", long_options, &option_index)) != -1) {
        switch (c) {
        case HELP_OPT:
            help();
            break;
        case VERSION_OPT:
            version();
            break;
        case UPDATE_OPT:
            action = UPDATE_OPT;
            break;
        case UPGRADE_OPT:
            action = UPGRADE_OPT;
            break;
        case LIST_OPT:
            action = LIST_OPT;
            break;
        case CLEAN_OPT:
            action = CLEAN_OPT;
            break;
        case FETCH_OPT:
            action = FETCH_OPT;
            slapt_vector_t_add(names, strdup(optarg));
            break;
        case SEARCH_OPT:
            action = SEARCH_OPT;
            slapt_vector_t_add(names, strdup(optarg));
            break;
        case SHOW_OPT:
            action = SHOW_OPT;
            slapt_vector_t_add(names, strdup(optarg));
            break;
        case BUILD_OPT:
            action = BUILD_OPT;
            slapt_vector_t_add(names, strdup(optarg));
            break;
        case INSTALL_OPT:
            action = INSTALL_OPT;
            slapt_vector_t_add(names, strdup(optarg));
            break;
        case YES_OPT:
            prompt = false;
            break;
        case SIMULATE_OPT:
            simulate = true;
            break;
        case NODEP_OPT:
            do_dep = false;
            break;
        case CONFIG_OPT:
            config_file = strdup(optarg);
            break;
        case POSTCMD_OPT:
            postcmd = strdup(optarg);
            break;
        case BUILD_ONLY_OPT:
            only_flags |= BUILD_ONLY_FLAG;
            break;
        case FETCH_ONLY_OPT:
            only_flags |= FETCH_ONLY_FLAG;
            break;
        default:
            help();
            exit(EXIT_FAILURE);
        }
    }

    if ((action == UPGRADE_OPT) && (optind < argc)) {
        fprintf(stderr, gettext("Individual packages not accepted when upgrading all slackbuilds\n"));
        exit(EXIT_FAILURE);
    }

    if ((only_flags & BUILD_ONLY_FLAG) && (only_flags & FETCH_ONLY_FLAG)) {
        fprintf(stderr, gettext("build-only and fetch-only options are mutually exclusive\n"));
        exit(EXIT_FAILURE);
    }

    /* add extra arguments */
    while (optind < argc) {
        slapt_vector_t_add(names, strdup(argv[optind]));
        ++optind;
    }

    slapt_src_config *config = NULL;
    if (config_file != NULL)
        config = slapt_src_read_config(config_file);
    else
        config = slapt_src_read_config(SLAPT_SRC_RC);
    if (config == NULL) {
        fprintf(stderr, gettext("Failed to read %s\n"), SLAPT_SRC_RC);
        exit(EXIT_FAILURE);
    }

    /* honor command line option */
    config->do_dep = do_dep;
    config->prompt = prompt;
    config->postcmd = postcmd; /* to be freed in slapt_src_config_free */

    init_builddir(config);
    if ((chdir(config->builddir)) != 0) {
        perror(gettext("Failed to chdir to build directory"));
        exit(EXIT_FAILURE);
    }

    slapt_vector_t *sbs = NULL;
    slapt_vector_t *remote_sbs = NULL;
    slapt_vector_t *installed = NULL;

    /* setup, fetch, and other preperation steps */
    switch (action) {
    case LIST_OPT:
    case SEARCH_OPT:
    case SHOW_OPT:
        remote_sbs = slapt_src_get_available_slackbuilds();
        break;
    case FETCH_OPT:
    case BUILD_OPT:
    case INSTALL_OPT:
    case UPGRADE_OPT:
        remote_sbs = slapt_src_get_available_slackbuilds();
        installed = slapt_get_installed_pkgs();
        /* convert all names to slackbuilds */
        if (names->size > 0) {
            sbs = slapt_src_names_to_slackbuilds(config, remote_sbs, names, installed);
            if (sbs == NULL || sbs->size == 0) {
                printf(gettext("Unable to find all specified slackbuilds.\n"));
                exit(EXIT_FAILURE);
            }
        } else if (action == UPGRADE_OPT) {
            /* for each entry in 'installed' see if it's available as a slackbuild */
            slapt_vector_t_foreach(const slapt_pkg_t *, pkg, installed) {
                slapt_vector_t *matches = slapt_vector_t_search(remote_sbs, sb_compare_pkg_to_name, pkg->name);
                if (!matches) {
                    continue;
                }
                slapt_vector_t_foreach(const slapt_src_slackbuild *, upgrade_sb, matches) {
                    if (slapt_pkg_t_cmp_versions(upgrade_sb->version, pkg->version) == 1) {
                        slapt_vector_t_add(names, strdup(upgrade_sb->name));
                    }
                }
                slapt_vector_t_free(matches);
            }

            sbs = slapt_src_names_to_slackbuilds(config, remote_sbs, names, installed);
        }
        /* provide summary */
        if (!simulate)
            action = show_summary(sbs, names, action, prompt);
        break;
    }

    /*
  ** When upgrade-all is specified the sbs list is automatically
  ** determined. Transform the action flag accord to any additional
  ** flags.
  */
    if (action == UPGRADE_OPT) {
        if (only_flags & BUILD_ONLY_FLAG) {
            action = BUILD_OPT;
        } else if (only_flags & FETCH_ONLY_FLAG) {
            action = FETCH_OPT;
        } else {
            action = INSTALL_OPT;
        }
    }

    /* now, actually do what was requested */
    switch (action) {
    case UPDATE_OPT:
        if (!slapt_src_update_slackbuild_cache(config))
            exit(EXIT_FAILURE);
        break;

    case FETCH_OPT:
        ;
        bool old_prompt = config->prompt;
        config->prompt = false;
        slapt_vector_t_foreach(slapt_src_slackbuild *, fetch_sb, sbs) {
            if (simulate) {
                printf(gettext("FETCH: %s\n"), fetch_sb->name);
                continue;
            } else
                slapt_src_fetch_slackbuild(config, fetch_sb);
        }
        config->prompt = old_prompt;
        break;

    case BUILD_OPT:
        ;
        slapt_vector_t_foreach(slapt_src_slackbuild *, build_sb, sbs) {
            const int nv_len = strlen(build_sb->name) + strlen(build_sb->version) + 2;
            char namever[nv_len];
            const int r = snprintf(namever, nv_len, "%s:%s", build_sb->name, build_sb->version);
            if (r + 1 != nv_len) {
                exit(EXIT_FAILURE);
            }

            if (simulate) {
                printf(gettext("BUILD: %s\n"), build_sb->name);
                continue;
            }

            if (!slapt_src_fetch_slackbuild(config, build_sb))
                exit(EXIT_FAILURE);
            slapt_src_build_slackbuild(config, build_sb);

            /* XXX we assume if we didn't request the slackbuild, then it is a dependency, and needs to be installed */
            slapt_vector_t *name_matches = slapt_vector_t_search(names, sb_compare_name_to_name, build_sb->name);
            slapt_vector_t *namever_matches = slapt_vector_t_search(names, sb_compare_name_to_name, namever);
            if (name_matches == NULL && namever_matches == NULL) {
                slapt_src_install_slackbuild(config, build_sb);
            }

            if (name_matches)
                slapt_vector_t_free(name_matches);
            if (namever_matches)
                slapt_vector_t_free(namever_matches);
        }
        break;

    case INSTALL_OPT:
        ;
        slapt_vector_t_foreach(slapt_src_slackbuild *, install_sb, sbs) {
            if (simulate) {
                printf(gettext("INSTALL: %s\n"), install_sb->name);
                continue;
            }

            if (!slapt_src_fetch_slackbuild(config, install_sb))
                exit(EXIT_FAILURE);
            slapt_src_build_slackbuild(config, install_sb);
            slapt_src_install_slackbuild(config, install_sb);
        }
        break;

    case SEARCH_OPT: {
        ;
        slapt_vector_t *search = slapt_src_search_slackbuild_cache(remote_sbs, names);
        slapt_vector_t_foreach(slapt_src_slackbuild *, search_sb, search) {
            printf("%s:%s - %s\n",
                   search_sb->name,
                   search_sb->version,
                   search_sb->short_desc != NULL ? search_sb->short_desc : "");
        }
        slapt_vector_t_free(search);
    } break;

    case LIST_OPT:
        ;
        slapt_vector_t_foreach(slapt_src_slackbuild *, list_sb, remote_sbs) {
            printf("%s:%s - %s\n",
                   list_sb->name,
                   list_sb->version,
                   list_sb->short_desc != NULL ? list_sb->short_desc : "");
        }
        break;

    case SHOW_OPT: {
        ;
        slapt_vector_t_foreach(const char *, show_name, names) {
            slapt_vector_t *parts = slapt_parse_delimited_list(show_name, ':');
            const char *name = parts->items[0];
            const char *ver = NULL;

            if (parts->size > 1)
                ver = parts->items[1];

            slapt_src_slackbuild *sb = slapt_src_get_slackbuild(remote_sbs, name, ver);

            if (sb != NULL) {
                printf(gettext("SlackBuild Name: %s\n"), sb->name);
                printf(gettext("SlackBuild Version: %s\n"), sb->version);
                printf(gettext("SlackBuild Category: %s\n"), sb->location);
                printf(gettext("SlackBuild Description: %s\n"), sb->short_desc != NULL ? sb->short_desc : "");

                printf(gettext("SlackBuild Files:\n"));

                slapt_vector_t_foreach(char *, f, sb->files) {
                    printf(" %s\n", f);
                }

                if (sb->requires != NULL)
                    printf(gettext("SlackBuild Requires: %s\n"), sb->requires);

                 printf("\n");

                /* slapt_src_slackbuild_free (sb); NO FREE */
            }
            slapt_vector_t_free(parts);
        }
    } break;
    case CLEAN_OPT:
        clean(config);
        break;
    }

    if (names != NULL)
        slapt_vector_t_free(names);
    if (sbs != NULL)
        slapt_vector_t_free(sbs);
    if (remote_sbs != NULL)
        slapt_vector_t_free(remote_sbs);
    if (installed != NULL)
        slapt_vector_t_free(installed);
    if (config_file != NULL)
        free(config_file);

    slapt_src_config_free(config);

    return 0;
}

static int show_summary(slapt_vector_t *sbs, slapt_vector_t *names, int action, bool prompt)
{
    int line_len = 0;

    if (sbs->size == 0) {
        printf(gettext("Done\n"));
        return action;
    }

    printf(gettext("The following packages will be %s:\n"),
           action == INSTALL_OPT ? gettext("installed")
                                 : action == BUILD_OPT ? gettext("built")
                                                       : gettext("fetched"));

    slapt_vector_t_foreach(const char *, sb_name, names) {
        slapt_vector_t *parts = slapt_parse_delimited_list(sb_name, ':');
        const char *name = parts->items[0];
        const int name_len = strlen(name);

        if (line_len == 0) {
            printf(" %s ", name);
            line_len += name_len + 1;
        } else if (line_len < 80 - name_len) {
            printf("%s ", name);
            line_len += name_len + 1;
        } else {
            printf("\n %s ", name);
            line_len = name_len + 2;
        }

        slapt_vector_t_free(parts);
    }
    printf("\n");

    if (names->size < sbs->size) {
        line_len = 0;

        printf(gettext("The following dependent slackbuilds will be built and installed:\n"));

        slapt_vector_t_foreach(const slapt_src_slackbuild *, sb, sbs) {
            const char *name = sb->name;
            const char *version = sb->version;
            char *namever = slapt_malloc(sizeof *namever * (strlen(name) + strlen(version) + 2));
            sprintf(namever, "%s:%s", name, version);

            slapt_vector_t *name_matches = slapt_vector_t_search(names, sb_compare_name_to_name, (char *)name);
            slapt_vector_t *namever_matches = slapt_vector_t_search(names, sb_compare_name_to_name, namever);
            if (name_matches == NULL && namever_matches == NULL) {
                const int name_len = strlen(name);

                if (line_len == 0) {
                    printf(" %s ", name);
                    line_len += name_len + 2;
                } else if (line_len < 80 - name_len) {
                    printf("%s ", name);
                    line_len += name_len + 1;
                } else {
                    printf("\n %s ", name);
                    line_len = name_len + 2;
                }
            }
            if (name_matches)
                slapt_vector_t_free(name_matches);
            if (namever_matches)
                slapt_vector_t_free(namever_matches);
            free(namever);
        }
        printf("\n");
    }

    if ((sbs->size > 0) && (prompt == true)) {
        if (slapt_ask_yes_no(gettext("Do you want to continue? [y/N] ")) != 1) {
            printf(gettext("Abort.\n"));
            return 0;
        }
    }

    return action;
}

static void clean(slapt_src_config *config)
{
    struct dirent *file = NULL;
    DIR *builddir = opendir(config->builddir);
    if (builddir != NULL) {
        while ((file = readdir(builddir)) != NULL) {
            struct stat stat_buf;

            if (strcmp(file->d_name, "..") == 0 || strcmp(file->d_name, ".") == 0)
                continue;

            if (lstat(file->d_name, &stat_buf) == -1)
                continue;

            if (S_ISDIR(stat_buf.st_mode)) {
                const int command_len = strlen(file->d_name) + 8;
                char command[command_len];
                const int r = snprintf(command, command_len, "rm -rf %s", file->d_name);
                if (r + 1 == command_len) {
                    int sys_r = system(command);
                    if (sys_r != 0)
                        exit(EXIT_FAILURE);
                }
            }
        }
        closedir(builddir);
    }
}
